course7/homeworks/fred/complex-rpn/utils.h

```diff
@@ -0,0 +1,66 @@
+#include "complex.h"
+#include "calculator.h"
+
+#include <string>
+#include <stdexcept>
+#include <utility>
+
+Complex parseComplexFromString(const std::string& s) {
+  std::string::size_type sz;
+
+  double realPart = std::stod(s, &sz);
+
+  double imaginaryPart = 0.0;
+  if (sz < s.size()) {
+    if (s[sz] == '+') {
+      ++sz;
+    }
+    else if (s[sz] != '-') {
+      throw std::runtime_error("Invalid complex number");
+    }
+
+    auto secondPart = s.substr(sz);
+    imaginaryPart = std::stod(secondPart, &sz);
+
+    if (sz != secondPart.size() - 1 || secondPart[sz] != 'i') {
+      throw std::runtime_error("Invalid complex number");
+    }
+  }
+
+  return Complex{ realPart, imaginaryPart };
+}
+
+std::pair<bool, Complex> tryParseComplexFromString(const std::string& s) {
+  try {
+    auto result = parseComplexFromString(s);
+    return std::make_pair(true, result);
+  }
```

Comment:
Throwing exceptions in C++ is expensive, and shouldn't be use as a part of normal logic flow.